test/test_torch.py

```diff
@@ -7065,6 +7065,7 @@ def test_matrix_exp_boundary_cases(self, device, dtype):
     @skipCUDAIfNoMagma
     @skipCPUIfNoLapack
     @dtypes(torch.float, torch.double)
+    @tf32_on_and_off(5.0)  # values are 40x.xx vs 40x.xx
```

Comment:
ouch! On the other hand, for matrix exp I'm not particularly surprised.

Reply:
So far we've disabled TF32 for linear algebra calls e.g. via MAGMA. Would it make sense to disable it for `matrix_exp` as well or at least raise a warning?

Reply:
According to my test, `matrix_exp` TF32 can be 6x faster than FP32. Different from magma, which is CPU bound, TF32 flags on `matrix_exp` do make a difference.

Reply:
Here is the test for precision:
```python
import torch

for i in range(5, 15):
    size = 2 ** i
    print(f"size: {size}x{size}")
    m1 = torch.randn(size, size, device='cuda')
    m2 = torch.randn(size, size, device='cuda')
    torch.backends.cuda.matmul.allow_tf32 = False
    e1 = m1 @ m2
    e2 = torch.matrix_exp(m1)
    torch.backends.cuda.matmul.allow_tf32 = True
    r1 = m1 @ m2
    r2 = torch.matrix_exp(m1)
    print("matmul error:", (r1 - e1).abs().max())
    print("exp error:", (r2 - e2).abs().max())
    print("matmul relative error:", (r1 - e1).abs().max() / e1.abs().max())
    print("exp relative error:", (r2 - e2).abs().max() / e2.abs().max())
    print()
```
```
size: 32x32
matmul error: tensor(3.8147e-06, device='cuda:0')
exp error: tensor(4.7684e-05, device='cuda:0')
matmul relative error: tensor(1.3857e-07, device='cuda:0')
exp relative error: tensor(5.0852e-07, device='cuda:0')

size: 64x64
matmul error: tensor(0.0096, device='cuda:0')
exp error: tensor(7.3298, device='cuda:0')
matmul relative error: tensor(0.0004, device='cuda:0')
exp relative error: tensor(0.0063, device='cuda:0')

size: 128x128
matmul error: tensor(0.0167, device='cuda:0')
exp error: tensor(183.8311, device='cuda:0')
matmul relative error: tensor(0.0004, device='cuda:0')
exp relative error: tensor(0.0105, device='cuda:0')

size: 256x256
matmul error: tensor(0.0212, device='cuda:0')
exp error: tensor(11040.5312, device='cuda:0')
matmul relative error: tensor(0.0003, device='cuda:0')
exp relative error: tensor(0.0102, device='cuda:0')

size: 512x512
matmul error: tensor(0.0325, device='cuda:0')
exp error: tensor(14844504., device='cuda:0')
matmul relative error: tensor(0.0003, device='cuda:0')
exp relative error: tensor(0.0215, device='cuda:0')

size: 1024x1024
matmul error: tensor(0.0432, device='cuda:0')
exp error: tensor(3.1523e+11, device='cuda:0')
matmul relative error: tensor(0.0003, device='cuda:0')
exp relative error: tensor(0.0482, device='cuda:0')

size: 2048x2048
matmul error: tensor(0.0715, device='cuda:0')
exp error: tensor(2.8972e+17, device='cuda:0')
matmul relative error: tensor(0.0003, device='cuda:0')
exp relative error: tensor(0.1296, device='cuda:0')

size: 4096x4096
matmul error: tensor(0.1049, device='cuda:0')
exp error: tensor(7.2534e+25, device='cuda:0')
matmul relative error: tensor(0.0003, device='cuda:0')
exp relative error: tensor(0.3424, device='cuda:0')

size: 8192x8192
matmul error: tensor(0.1492, device='cuda:0')
exp error: tensor(2.3980e+37, device='cuda:0')
matmul relative error: tensor(0.0003, device='cuda:0')
exp relative error: tensor(0.5230, device='cuda:0')

size: 16384x16384
matmul error: tensor(0.2262, device='cuda:0')
exp error: tensor(nan, device='cuda:0')
matmul relative error: tensor(0.0003, device='cuda:0')
exp relative error: tensor(nan, device='cuda:0')

```

Reply:
I don't think there is an easy way to disable TF32 for `matrix_exp`. It is implemented using other aten operators, and `at::matmul` is where the TF32 flag play a role. You can not just disable this flag at the beginning of `matrix_exp`, because these flags are not threadsafe.

Reply:
Maybe we should just issue a warning:

```
You are computing matrix_exp with allow_tf32=True. With this flag, the speed of computation
can be up to ~6x faster, but the precision could be much worse. 

To learn more about TF32, check:
https://pytorch.org/docs/master/notes/cuda.html#tensorfloat-32-tf32-on-ampere-devices

To run matrix_exp without TF32, do:
torch.backends.cuda.matmul.allow_tf32 = False
result = torch.matrix_exp(mymat)
torch.backends.cuda.matmul.allow_tf32 = True
```

Reply:
@ngimel @ptrblck Any opinion?

Reply:
@ngimel Can we not acquire the current cuBLAS handle and use the above guard before calling matrix_exp? As long as the functions matrix_exp is composed of don't use multiple cuBLAS handles maybe that will work?

`matrix_exp` appears to be very little used so we'd prefer to avoid a precision headache. If it's difficult to call it with tf32 disabled locally then throwing a warning seems OK.

Reply:
Are there other cases of poor precision that aren't covered yet? 

@ngimel suggests all linear algebra computations other than matmul should disable tf32. Most of these operations use MAGMA currently, but this PR is going to use cuSolver and cuBLAS calls https://github.com/pytorch/pytorch/pull/42403. Do we need to update it or put warnings around these operations for how to handle tf32?

Reply:
I have disabled TF32 on `matrix_exp` by introducing a `thread_local` variable to override the TF32 flag.

Reply:
Can we deal with cusolver ops in later PR? I don't know which op will be affected now. After @xwang233 has migrated ops to cusolver, I will run tests to see if it is needed to disable TF32.